core/cap-tbd.md

```diff
@@ -0,0 +1,209 @@
+
+
+### Data Types
+Following types are used throughout this proposal.
+
+```
+ const BLST_G1_POINT: usize = 96; (rename after adding decoding functions)
+ const BLST_G2_POINT: usize = 192;
+ const BLST_SCALAR_SIZE: usize = 255;
+```
+### Curve Operations
+
+#### Addition in G1
+
+```rust
+fn bls12_g1add( e: &Host, p0: BLST_G1_POINT, p1: BLST_G1_POINT) -> Result <BLST_G1_POINT, Error>;
+```
+Perform addition of two points in G1. 
+Inputs should be 96 bytes each. 
+An error will be thrown if neither input is in G1, input has invalid length, or encoding rules are violated. 
+
+#### Multiplication in G1
+
+```rust
+fn bls12_g1mul( e: &Host, scalar: u8, p0: BLST_G1_POINT) -> Result <BLST_G1_POINT, Error>;
+```
+Perform multiplication between a point in G1 and a scalar.
+Input should be 96 bytes and 8 bytes.
+An error will be thrown if the input p0 is not in G1, input p1 has invalid length, or encoding rules are violated. 
+
+#### Multiexponentiation in G1
+
+```rust
+fn bls12_g1multiexp( e: &Host, scalars: &[u8], p_n: BLST_G1_POINT) -> Result <BLST_G1_POINT, Error>;
+```
+Perform multipliexponentiation on a point in G1 and a scalar. 
+Input should be 96 bytes and 8 bytes.
+An error will be thrown if the input p_n is not in G1, input p1 has invalid length, or encoding rules are violated. 
+
+#### Addition in G2
+
+```rust
+fn bls12_g2add( e: &Host, p0: BLST_G2_POINT, p1: BLST_G2_POINT) -> Result <BLST_G2_POINT, Error>;
+```
+Perform addition of two points in G2. 
+Inputs should be 192 bytes each. 
+An error will be thrown if neither input is in G2, input has invalid length, or encoding rules are violated. 
+
+#### Multiplication in G2
+
+```rust
+fn bls12_g2mul( e: &Host, scalar: u8, p0: BLST_G2_POINT) -> Result <BLST_G2_POINT, Error>;
+```
+Perform multiplication between a point in G2 and a scalar.
+Input should be 192 bytes and 8 bytes.
+An error will be thrown if the input p0 is not in G2, input p2 has invalid length, or encoding rules are violated. 
+
+#### Multiexponentiation in G2
+
+```rust
+fn bls12_g2multiexp( e: &Host, scalars: &[u8], p_n: BLST_G2_POINT) -> Result <BLST_G2_POINT, Error>;
+```
+Perform multipliexponentiation on a point in G2 and a scalar. 
+Input should be 192 bytes and 8 bytes.
+An error will be thrown if the input p_n is not in G2, input p_n has invalid length, or encoding rules are violated. 
+
+#### Pairing
+
+```rust
+fn bls12_pairing( e: &Host, p_g1: &[BLST_G1_POINT], p_g2: &[BLST_G2_POINT]) -> Result <u32, Error>;
+```
+
+Perform pairing between points in G1 and G2. 
+Arguments: 
+p_g1: point in G1(96 bytes)
+p_g1: point in G1(192 bytes)
+Returns padded 0x01 if pairing is successful else 0x00.
+Error will be thrown if p0_g1 and p1_g1 are not on G1 and G2 respectively, if the points are in the incorrect subgroups, have invalid length, inputs are empty, encoding rules are violated. 
+
+#### Map Base Field Element to G1
+
+```rust
+fn bls12_map_fp_to_g1( e: &Host, fp: &[u64]) -> Result <BLST_G1_POINT, Error>;
```

Comment:
Why do we have two variants of the same 'map' function that seemingly take the same input? I think we should only need one version, that either takes a `U64Val`, or `BytesObject`, depending on which case is more useful.